model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentSessionsSingleUseObjectProviderFactory.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.sessions.infinispan;
+
+
+import org.infinispan.commons.api.BasicCache;
+import org.keycloak.Config;
+import org.keycloak.common.Profile;
+import org.keycloak.common.util.Time;
+import org.keycloak.connections.infinispan.InfinispanUtil;
+import org.keycloak.infinispan.util.InfinispanUtils;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.SingleUseObjectProvider;
+import org.keycloak.models.session.RevokedTokenPersisterProvider;
+import org.keycloak.models.sessions.infinispan.entities.SingleUseObjectValueEntity;
+import org.keycloak.models.utils.PostMigrationEvent;
+import org.keycloak.provider.ProviderEvent;
+import org.keycloak.provider.ProviderEventListener;
+import org.keycloak.services.scheduled.ClearExpiredRevokedTokens;
+import org.keycloak.services.scheduled.ClusterAwareScheduledTaskRunner;
+import org.keycloak.timer.TimerProvider;
+
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Extends the {@link InfinispanSingleUseObjectProviderFactory} to read expired tokens from the database on startup if no other node has loaded the data already.
+ * @author Alexander Schwartz
+ */
+public class PersistentSessionsSingleUseObjectProviderFactory extends InfinispanSingleUseObjectProviderFactory {
+
+    volatile boolean initialized;
+
+    private final static String LOADED = "loaded" + SingleUseObjectProvider.REVOKED_KEY;
+
+    @Override
+    public PersistentSessionsSingleUseObjectProvider create(KeycloakSession session) {
+        initialize(session);
+        return new PersistentSessionsSingleUseObjectProvider(session, singleUseObjectCache);
+    }
+
+    private void initialize(KeycloakSession session) {
+        if (!initialized) {
+            synchronized (this) {
+                if (!initialized) {
+                    RevokedTokenPersisterProvider provider = session.getProvider(RevokedTokenPersisterProvider.class);
+                    BasicCache<String, SingleUseObjectValueEntity> cache = singleUseObjectCache.get();
```

Comment:
Maybe add a `provider != null` check here, because then it would be possible to disable this by simply disabling the JpaRevokedTokensPersister provider?